src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+        //console.log("mono_wasm_invoke_js_1 E");
+        return InvokeJSResult.Success;
+    } catch (ex) {
+        //console.log("mono_wasm_invoke_js_1 F");
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+export function mono_wasm_bind_js_function(interned_function_id: MonoString, arg_count: number, signature: VoidPtr, function_js_handle: Int32Ptr, is_exception: Int32Ptr): MonoObject {
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        const closure: any = { MONO, BINDING, INTERNAL, fn };
+        let body = "try {\n";
+
+        for (let position = 1; position <= arg_count; position++) {
+            const offset = position << 2;
+            const mono_type = <MonoType><any>getU32(<any>signature + offset);
+            const covertor = cs_to_js_map.get(mono_type);
+            assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+            const convertor_name = `covertor${position}`;
+            closure[convertor_name] = covertor;
+
+            body += `  const arg${position}_js = closure.${convertor_name}(<any>data + ${offset});\n`;
+        }
+        if (arg_count === 0) {
+            body += "  const res_js = closure.fn();\n";
+        }
+        else if (arg_count === 1) {
+            body += "  const res_js = closure.fn(arg1_js);\n";
+        }
+
+        const res_mono_type = <MonoType><any>getU32(<any>signature);
+        if (res_mono_type === MonoTypeNull) {
+            body += "  assert(value === undefined, 'Function returned unexpected value, C# signature is void');\n";
+        } else {
+            const covertor = js_to_cs_map.get(res_mono_type);
+            assert(covertor, `Unknow covertor for type ${res_mono_type} at ret`);
+            closure["covertor0"] = covertor;
+
+            body += "  closure.covertor0(<any>data);\n";
+        }
+
+        body += "  return InvokeJSResult.Success;\n";
+        body += "} catch (ex) {\n";
+        body += "  _marshal_exception_to_cs(data, ex);\n";
+        body += "  return InvokeJSResult.FunctionThrewException;\n";
+        body += "}\n";
+
+        const bound_fn = new Function("closure", "data", body).bind(null, closure);
+        const bound_function_handle = mono_wasm_get_js_handle(bound_fn)!;
+        setI32(function_js_handle, <any>bound_function_handle);
+        return MonoObjectNull;
+    } catch (ex) {
+        return wrap_error(is_exception, ex);
+    }
+}
+
+export function mono_wasm_invoke_bound_function(function_js_handle: JSHandle, data: VoidPtr): InvokeJSResult {
+    const js_fn = mono_wasm_get_jsobj_from_js_handle(function_js_handle);
+    assert(js_fn, `Unknow function for handle ${function_js_handle}`);
+    return js_fn(data);
+}
+
+
+
+function mono_wasm_lookup_function(function_id: MonoString): Function {
+    let res = _invoke_js_function_cache.get(function_id);
+    if (res !== undefined) {
```

Comment:
We need to instead check whether the cache ```has``` the id, and if so unconditionally return the result. Otherwise we won't be caching our first lookup result in all cases and will walk the global scope again Sometimes, when we want to never walk it more than once.

Reply:
Sorry, I don't follow. That's what it does, no ?

Reply:
Are you saying that we should cache flattened object nesting hierarchy ?

Reply:
The first lookup may find 'undefined' because the function they tried to locate doesn't exist, so we should cache that to maintain consistent behavior (that we always cache the first result we get).

Reply:
Why is that worth two hash table operations `has` and `get` on each call ?

I actually removed the same pattern in the string interning code, not knowing why it's there.
There it's more performance critical thought.

Reply:
If we're OK with breaking the behavior for performance, that's probably an okay decision, we just need to document it. Isn't the performance critical path going to require you to bind the method once in advance, making the cost of correctness not significant?

Reply:
"Correctness" is what I'm asking about. I don't know why caching failure is more correct than trying again next time. 
For this function, I think nobody would try to call it multiple times with same name anyway. If it failed for the first time, they will probable stop the application.